File: frontend/src/types/focus-session.ts
export type PlannedMinutes = 25 | 50 | 90;

export interface FocusSession {
  id: number;
  user_id: number;
  task_id: number | null;
  plan_item_id: number | null;
  started_at: string;
  ended_at: string | null;
  planned_minutes: number;
  auto_closed: boolean;
  actual_minutes: number | null;
  task_title: string | null;
  plan_item_title: string | null;
}

export interface StartFocusBody {
  task_id?: number | null;
  plan_item_id?: number | null;
  planned_minutes: PlannedMinutes;
}

export interface FocusSessionTodayResponse {
  sessions: FocusSession[];
  total_minutes: number;
  count: number;
}

/**
 * Format minutes as compact human-readable label.
 * Examples: 0 → "0m", 45 → "45m", 60 → "1h", 135 → "2h 15m".
 */
export function formatFocusMinutes(mins: number): string {
  if (mins < 60) return `${mins}m`;
  const h = Math.floor(mins / 60);
  const m = mins % 60;
  return m === 0 ? `${h}h` : `${h}h ${m}m`;
}

File: backend/app/schemas/focus_session.py
"""Pydantic schemas for the focus_sessions API (D12).

``FocusSessionResponse`` uses ``model_config = ConfigDict(from_attributes=True)``
so the service layer can build it directly from a SQLAlchemy ORM instance.
The ``actual_minutes`` field is derived by the service from
``started_at``/``ended_at`` (not a stored column).
"""
from __future__ import annotations

from datetime import datetime
from typing import Literal, Optional

from pydantic import BaseModel, ConfigDict, Field


PlannedMinutes = Literal[25, 50, 90]


class FocusSessionStart(BaseModel):
    action_id: Optional[int] = None
    plan_item_id: Optional[int] = None
    planned_minutes: PlannedMinutes


class FocusSessionResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    user_id: int
    action_id: Optional[int] = None
    plan_item_id: Optional[int] = None
    started_at: datetime
    ended_at: Optional[datetime] = None
    planned_minutes: int
    auto_closed: bool
    actual_minutes: Optional[int] = None
    action_title: Optional[str] = None
    plan_item_title: Optional[str] = None


class FocusSessionTodayResponse(BaseModel):
    sessions: list[FocusSessionResponse]
    total_minutes: int = Field(ge=0)
    count: int = Field(ge=0)
